In [2]:
%run run_demo.py

ticker: AAPL


[*********************100%***********************]  1 of 1 completed
/Users/sambrown04/Documents/570/financial-reasoning-llm/src/anomaly_detection.py:18: FutureWarning: Calling float on a single element Series is deprecated and will raise a TypeError in the future. Use float(ser.iloc[0]) instead
  "move": float(row["return_pct"]),


found 7 big moves
Starting news pull
pulling news from 2024-05-03
pulling news from 2024-06-11
pulling news from 2025-04-03
pulling news from 2025-04-04
pulling news from 2025-04-09
pulling news from 2025-05-12
pulling news from 2025-08-06
2024-05-03 (5.98%): 50 headlines
2024-06-11 (7.26%): 50 headlines
2025-04-03 (-9.25%): 50 headlines
2025-04-04 (-7.29%): 50 headlines
2025-04-09 (15.33%): 50 headlines
2025-05-12 (6.31%): 50 headlines
2025-08-06 (5.09%): 50 headlines

LLM explanation:

### Analysis of AAPL Price Moves and Related News

#### Date: 2024-05-03 | Move: +5.98%
**Key Headlines:**
- Warren Buffett's Berkshire Hathaway reduced its stake in Apple.
- Apple reported strong Q2 earnings, beating estimates.
- Analysts noted a potential AI-driven upgrade cycle for Apple products.

**Explanation:**
The significant price increase likely reflects investor optimism following Apple's strong earnings report, which indicated better-than-expected performance, particularly in iPhone sales. 

### Running the same call with my own API token

ticker: AAPL

[*********************100%***********************]  1 of 1 completed
finding big moves...
c:\Users\smant\financial-reasoning-llm\src\anomaly_detection.py:17: FutureWarning: Calling float on a single element Series is deprecated and will raise a TypeError in the future. Use float(ser.iloc[0]) instead
  "move": float(row["return_pct"]),
found 37 big moves

getting news from finviz...
recent headlines:
- Apple AI chief leaving as iPhone maker plays catch-up
- Apple's potential Intel partnership hinges on 18A-P process maturity by 2027
- Samsungs Next Salvo Against Apple: A Triple-Folding Smartphone
- Apple to Revamp AI Team After Announcing Top Executives Departure
- Here Are Billionaire Warren Buffett's 5 Biggest Stock Holdings
- Siri-us setback: Apples AI chief steps down as company lags behind rivals
- The Dealmaking 3  Paramount Lands UEFA Media Rights, Sony & ATP Medias Strategic Partnership, Apple Adds MLS
- A Retirement Marks Another Chapter in Apple's AI Reshuffle
- Apple AI chief Giannandrea to step down as company reshuffles leadership

asking LLM for explanation...

LLM explanation:

The recent price movements of Apple (AAPL) stock show a lot of volatility, with significant ups and downs. For example, there were sharp declines of over 9% in early April 2025, followed by a strong rebound of over 15% shortly after. This suggests that the stock is reacting strongly to news and market sentiment.

The headlines indicate that Apple is facing challenges, particularly in the AI space, as their AI chief is leaving and the company is trying to catch up with competitors. This leadership change could create uncertainty among investors, leading to stock price drops. Additionally, the mention of potential partnerships and competition from Samsung suggests that Apple is navigating a competitive landscape, which can also impact investor confidence.

Overall, the combination of leadership changes, competitive pressures, and market reactions to these developments may be contributing to the stock's volatility. Investors might be cautious about the company's direction, especially in AI, which is a critical area for future growth.

### Running llm_reasoning_playground with max_tokens = 1000
getting news from finviz...
recent headlines:
- Apple AI chief leaving as iPhone maker plays catch-up
- Apple's potential Intel partnership hinges on 18A-P process maturity by 2027
- Samsungs Next Salvo Against Apple: A Triple-Folding Smartphone
- Apple to Revamp AI Team After Announcing Top Executives Departure
- Here Are Billionaire Warren Buffett's 5 Biggest Stock Holdings
- Siri-us setback: Apples AI chief steps down as company lags behind rivals
- The Dealmaking 3  Paramount Lands UEFA Media Rights, Sony & ATP Medias Strategic Partnership, Apple Adds MLS
- A Retirement Marks Another Chapter in Apple's AI Reshuffle
- Apple AI chief Giannandrea to step down as company reshuffles leadership

asking LLM for explanation...

LLM explanation:

The stock market has experienced notable volatility recently, and Apple Inc. (AAPL) has mirrored this trend with significant price fluctuations. Over the past year, AAPL has shown a pattern of sharp increases followed by steep declines, indicating a high level of uncertainty and investor reaction to both company-specific news and broader market conditions.

AAPL's price changes reveal a series of ups and downs. For instance, after a drop of 3.58% at the beginning of January 2024, the stock rebounded with a 3.26% increase later that month. This pattern continued, with notable peaks such as a 7.26% increase in June 2024 and a staggering 15.33% rise in early April 2025, followed by significant drops, including a 9.25% decrease shortly after. This volatility suggests that investor sentiment is highly reactive to news and developments surrounding the company.

Recent headlines indicate that AAPL is facing challenges in the competitive landscape, particularly in artificial intelligence and smartphone technology. The departure of key AI executives has raised concerns about Apple's ability to keep pace with rivals, which may have contributed to the stock's downward movements. Additionally, the potential partnership with Intel and competition from Samsung's innovative products could be influencing investor confidence.

In the broader stock market context, fluctuations in AAPL's stock price may also be linked to general economic conditions, such as interest rate changes, inflation concerns, and market sentiment towards technology stocks. The tech sector has been particularly sensitive to macroeconomic factors, which can lead to increased volatility in stocks like AAPL.

In summary, AAPL's stock has experienced significant price changes driven by both internal challenges and external market conditions. The company's ongoing adjustments in leadership and strategy, particularly in AI, alongside competitive pressures, are key factors influencing its stock performance. Investors should remain cautious as the market continues to navigate these uncertainties.

In [ ]:
import json
import os
from datetime import datetime

def log_llm_output(model, temperature, max_tokens, output, ticker=None, log_file="llm_output_log.json"):
    # Create log entry
    log_entry = {
        "timestamp": datetime.now().isoformat(),
        "model": model,
        "temperature": temperature,
        "max_tokens": max_tokens,
        "ticker": ticker,
        "output": output,
        "output_length": len(output)
    }
    
    # Load existing logs or create new list
    logs = []
    if os.path.exists(log_file):
        try:
            with open(log_file, 'r') as f:
                logs = json.load(f)
        except (json.JSONDecodeError, IOError):
            logs = []
    
    # Append new entry
    logs.append(log_entry)
    
    # Write back to file
    with open(log_file, 'w') as f:
        json.dump(logs, f, indent=2)
    
    print(f"Logged output to {log_file}")
    return log_entry


def run_and_log(log_file="llm_output_log.json"):
    from run_demo import main as run_demo_main
    
    # Run the demo
    results = run_demo_main()
    
    if results is None:
        print("Demo failed - no results to log")
        return None
    
    # Extract data and log automatically
    config = results.get("llm_config", {})
    log_llm_output(
        model=config.get("model"),
        temperature=config.get("temperature"),
        max_tokens=config.get("max_tokens"),
        output=results.get("explanation"),
        ticker=results.get("ticker"),
        log_file=log_file
    )
    
    return results